src/platform/vita/vita_keyboard.c

```diff
@@ -0,0 +1,173 @@
+#include <string.h>
+#include <stdbool.h>
+#include <psp2/apputil.h>
+#include <psp2/display.h>
+#include <psp2/kernel/processmgr.h>
+#include <psp2/ime_dialog.h>
+#include <psp2/message_dialog.h>
+#include <vita2d.h>
+#include "vita_keyboard.h"
+
+#define IME_DIALOG_RESULT_NONE 0
+#define IME_DIALOG_RESULT_RUNNING 1
+#define IME_DIALOG_RESULT_FINISHED 2
+#define IME_DIALOG_RESULT_CANCELED 3
+
+static int ime_dialog_running = 0;
+static int ime_dialog_option = 0;
+static int ime_init_apputils = 0;
+
+static uint16_t ime_title_utf16[SCE_IME_DIALOG_MAX_TITLE_LENGTH];
+static uint16_t ime_initial_text_utf16[SCE_IME_DIALOG_MAX_TEXT_LENGTH];
+static uint16_t ime_input_text_utf16[SCE_IME_DIALOG_MAX_TEXT_LENGTH + 1];
+static uint8_t ime_input_text_utf8[SCE_IME_DIALOG_MAX_TEXT_LENGTH + 1];
+
+static void utf16_to_utf8(uint16_t *src, uint8_t *dst)
+{
+    int i;
+    for (i = 0; src[i]; i++) {
+        if ((src[i] & 0xFF80) == 0) {
+            *(dst++) = src[i] & 0xFF;
+        } else if((src[i] & 0xF800) == 0) {
+            *(dst++) = ((src[i] >> 6) & 0xFF) | 0xC0;
+            *(dst++) = (src[i] & 0x3F) | 0x80;
+        } else if((src[i] & 0xFC00) == 0xD800 && (src[i + 1] & 0xFC00) == 0xDC00) {
+            *(dst++) = (((src[i] + 64) >> 8) & 0x3) | 0xF0;
+            *(dst++) = (((src[i] >> 2) + 16) & 0x3F) | 0x80;
+            *(dst++) = ((src[i] >> 4) & 0x30) | 0x80 | ((src[i + 1] << 2) & 0xF);
+            *(dst++) = (src[i + 1] & 0x3F) | 0x80;
+            i += 1;
+        } else {
+            *(dst++) = ((src[i] >> 12) & 0xF) | 0xE0;
+            *(dst++) = ((src[i] >> 6) & 0x3F) | 0x80;
+            *(dst++) = (src[i] & 0x3F) | 0x80;
+        }
+    }
+
+    *dst = '\0';
+}
+
+static void utf8_to_utf16(uint8_t *src, uint16_t *dst)
+{
+    int i;
+    for (i = 0; src[i];) {
+        if ((src[i] & 0xE0) == 0xE0) {
+            *(dst++) = ((src[i] & 0x0F) << 12) | ((src[i + 1] & 0x3F) << 6) | (src[i + 2] & 0x3F);
+            i += 3;
+        } else if ((src[i] & 0xC0) == 0xC0) {
+            *(dst++) = ((src[i] & 0x1F) << 6) | (src[i + 1] & 0x3F);
+            i += 2;
+        } else {
+            *(dst++) = src[i];
+            i += 1;
+        }
+    }
+
+    *dst = '\0';
+}
+
+static int init_ime_dialog(char *title, const char *initial_text, int max_text_length, int type, int option)
+{
+    if (ime_dialog_running)
+    return -1;
```

Comment:
Indentation.

Reply:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb